app.go

```diff
@@ -413,92 +410,178 @@ func Serve() {
-	shttp.Handle("/", fs)
-	r.PathPrefix("/").Handler(fs)
+// CreateConfig creates a default configuration and saves it to the app's cfgFile.
+func CreateConfig(app *App) error {
+	log.Info("Creating configuration...")
+	c := config.New()
+	log.Info("Saving configuration %s...", app.cfgFile)
+	err := config.Save(c, app.cfgFile)
+	if err != nil {
+		return fmt.Errorf("Unable to save configuration: %v", err)
+	}
+	return nil
+}
 
-	// Handle shutdown
-	c := make(chan os.Signal, 2)
-	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
-	go func() {
-		<-c
-		log.Info("Shutting down...")
-		shutdown(app)
-		log.Info("Done.")
-		os.Exit(0)
-	}()
+// DoConfig runs the interactive configuration process.
+func DoConfig(app *App) {
+	d, err := config.Configure(app.cfgFile)
+	if err != nil {
+		log.Error("Unable to configure: %v", err)
+		os.Exit(1)
+	}
+	if d.User != nil {
+		app.cfg = d.Config
+		connectToDatabase(app)
+		defer shutdown(app)
 
-	http.Handle("/", r)
+		if !app.db.DatabaseInitialized() {
+			err = adminInitDatabase(app)
+			if err != nil {
+				log.Error(err.Error())
+				os.Exit(1)
+			}
+		}
 
-	// Start web application server
-	var bindAddress = app.cfg.Server.Bind
-	if bindAddress == "" {
-		bindAddress = "localhost"
+		u := &User{
+			Username:   d.User.Username,
+			HashedPass: d.User.HashedPass,
+			Created:    time.Now().Truncate(time.Second).UTC(),
+		}
+
+		// Create blog
+		log.Info("Creating user %s...\n", u.Username)
+		err = app.db.CreateUser(u, app.cfg.App.SiteName)
+		if err != nil {
+			log.Error("Unable to create user: %s", err)
+			os.Exit(1)
+		}
+		log.Info("Done!")
 	}
-	if app.cfg.IsSecureStandalone() {
-		log.Info("Serving redirects on http://%s:80", bindAddress)
-		go func() {
-			err = http.ListenAndServe(
-				fmt.Sprintf("%s:80", bindAddress), http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-					http.Redirect(w, r, app.cfg.App.Host, http.StatusMovedPermanently)
-				}))
-			log.Error("Unable to start redirect server: %v", err)
-		}()
+	os.Exit(0)
+}
 
-		log.Info("Serving on https://%s:443", bindAddress)
-		log.Info("---")
-		err = http.ListenAndServeTLS(
-			fmt.Sprintf("%s:443", bindAddress), app.cfg.Server.TLSCertPath, app.cfg.Server.TLSKeyPath, nil)
-	} else {
-		log.Info("Serving on http://%s:%d\n", bindAddress, app.cfg.Server.Port)
-		log.Info("---")
-		err = http.ListenAndServe(fmt.Sprintf("%s:%d", bindAddress, app.cfg.Server.Port), nil)
+// GenerateKeyFiles creates app encryption keys and saves them into the configured KeysParentDir.
+func GenerateKeyFiles(app *App) error {
+	// Read keys path from config
+	app.LoadConfig()
+
+	// Create keys dir if it doesn't exist yet
+	fullKeysDir := filepath.Join(app.cfg.Server.KeysParentDir, keysDir)
+	if _, err := os.Stat(fullKeysDir); os.IsNotExist(err) {
+		err = os.Mkdir(fullKeysDir, 0700)
+		if err != nil {
+			return err
+		}
 	}
+
+	// Generate keys
+	initKeyPaths(app)
+	var keyErrs error
```

Comment:
Should we use a multierror here? Otherwise we would only report as many as the last of three errors back to the caller.

Reply:
Possibly.. what's a multierror? I'm not familiar with it.

Reply:
It's a hashicorp library, we could use or implement something simliar. not a msut have but it's nice. hashicorp/go-multierror

Reply:
note: if you go with that don't forget `keys/keys.go`

Reply:
:+1: thanks, that could definitely be the way to go. I added a TODO in both places so we can be sure to tackle this in another iteration.